Initial Service Manager v4 Support

gsteel opened the pull request:
Provides initial support for SMv4

There is still a fair bit of work to do where the FilterPluginManager is consumed by FilterChain + other filters. These will need to be refactored with the FilterPluginManager as a hard-dependency in the constructor, otherwise none of the aliases will work now they have been extracted to the config provider.

Closes #125

gsteel committed:
Remove alias for Zend PluginManager

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Add default factory configuration for the standard filters in Laminas\Filter namespace

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
- Remove auto-wired aliases and factories
- Drop inclusion of i18n filters
- Drop legacy SMv2 validation

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Fix up MVC Module.php so that it returns filter configuration

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Upgrade to SMv4 and drop i18n and uri in dev

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Re-introduce legacy filter aliases to the config provider

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Use `AbstractPluginManager::build()` when a filter instantiation requires options.

`AbstractPluginManager::get()` no longer has an `options` parameter.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Bump dependencies

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve plugin manager doc blocks and add explicit, hard-coded configuration variables from abstract plugin manager

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Fix factory test…

- Remove fragile and now incorrect examination if plugin manager internals
- Fix configuration where an alias will no longer resolve because the plugin manager does not self-define all known filters

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Fix plugin manager (hidden dependency) instantiation inside FilterChain.

Wires up factories and aliases as configured in the package ConfigProvider so that the created plugin manager is functional.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Re-work the compatibility test copying tests from the trait shipped in service manager.

The test trait shipped on service manager cannot be used because it is designed for `SingleInstancePluginManager` which `FilterPluginManager` cannot inherit from.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Static analysis fixes and _some_ native types for FilterChain

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Reset baseline

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
CS Fixes

Signed-off-by: George Steel <george@net-glue.co.uk>

boesing commented on src/Module.php:
```diff
@@ -39,4 +44,10 @@ public function init($moduleManager)
+    /** @return ServiceManagerConfiguration */
+    public function getFilterConfig(): array
+    {
+        return (new ConfigProvider())()['filters'];
```
I'd rather provide a public method in `ConfigProvider` than fetching the array key but thats most probably just my personal preference.

froschdesign replied:
> I'd rather provide a public method in `ConfigProvider`…

This makes sense, because it is also used elsewhere:

```php
public function getPluginManager(): FilterPluginManager
{
    $plugins = $this->plugins;
    if (! $plugins instanceof FilterPluginManager) {
        $plugins = new FilterPluginManager(new ServiceManager(), (new ConfigProvider())()['filters']);
        $this->setPluginManager($plugins);
    }

    return $plugins;
}
```

https://github.com/laminas/laminas-filter/pull/136/files#diff-3a6c9663e22963eb296b993e0b7cc8335c9f7f02f8364fc66385ed7175b5639dR124

And also in tests.

gsteel replied:
Yes - referencing array keys is like a hidden public API and worse than simply exposing a public method 👍

gsteel replied:
I've removed filter config completely from the config provider in favour of a class constant on the plugin manager, as per `laminas-serializer` thanks to @boesing

boesing commented on src/ConfigProvider.php:
```diff
@@ -4,42 +4,244 @@
     {
         return [
             'dependencies' => $this->getDependencyConfig(),
+            'filters'      => $this->filters(),
```
I explicitly kept most of the plugin manager configuration within the plugin manager when approaching migration to SM v4.
That was to prevent more BC breaks than necessary.
As you can see, you already had to write that `CreatePluginManager` helper class to instantiate an instance of the plugin manager because of this.
I do not see a real benefit by having this config in the `ConfigProvider`. 
Only reason I have this in `laminas-cache` is due to the fact that we do have satellites there, so I'd rather move this back into the plugin manager instead.

gsteel replied:
My reasoning for moving config out of the plugin manager was that all of those protected props for factories, aliases, etc are now deprecated - it made sense to me to deal with deprecations at the same time considering we'll be releasing a major.

My plan in future PRs is to introduce factories for filters that rely on the plugin manager such as `FilterChain`, where the plugin manager will be made a constructor dependency.

Do you think the best course of action is to revert or to continue with that plan?

boesing replied:
Thats why I used config constructor Argument instead. you can see that in the serializer component for example

gsteel replied:
Thanks for the pointer - much better solution 👍

boesing reviewed (changes requested):
LGTM. Like to see first plugin manager to not implement the `AbstractSingleInstancePluginManager` class as we do have to allow `callable` and `FilterInterface` types.

However, I would keep the plugin manager configuration in the plugin manager to prevent too much of a BC break. Especially a break which is not really necessary IMO.

boesing commented:
I wonder if we should provide a `CallableFilter` with v4 so that we do not have this `mixed` stuff laying around.
Just not 100% sure where and how to map callables to the `CallableFilter` but that would at least reduce complexity a little bit. Just some thought which came to my mind, nothing we have to do - just wanted to drop that here.

gsteel commented:
> I wonder if we should provide a `CallableFilter` with v4 so that we do not have this `mixed` stuff laying around.
> Just not 100% sure where and how to map callables to the `CallableFilter` but that would at least reduce complexity a little bit. Just some thought which came to my mind, nothing we have to do - just wanted to drop that here.

I was thinking the same thing. Probably worth exploring in a separate patch

gsteel committed:
Set default filter config as a constant on the plugin manager

To reduce BC breaks, retain the concept of default configuration of the plugin manager, moving configuration out of the ConfigProvider and into a class constant.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
@boesing - I'd appreciate another review when you get a chance! Cheers.

boesing commented:
will do later, have to take a 4h train and thus need some stuff to do 😁

boesing commented on src/FilterChain.php:
```diff
@@ -189,19 +180,13 @@ public function attach($callback, $priority = self::DEFAULT_PRIORITY)
      * with the retrieved instance.
      *
-     * @param  string $name
+     * @param class-string|string $name
```
I'd say this should be class-string which matches `InstanceType`, no?

boesing commented on test/FilterPluginManagerCompatibilityTest.php:
```diff
@@ -5,53 +5,84 @@
 
 class FilterPluginManagerCompatibilityTest extends TestCase
 {
-    use CommonPluginManagerTrait;
```
did we drop this on purpose? is that due to the `callable` type?

gsteel replied:
Yes, and also that trait only works with `SingleInstancePluginManager`

boesing reviewed (approved):
LGTM, just a few questions but overall pretty neat work.

gsteel committed:
Refine argument to `attachByName` to limit possible class-string args to descendants of `FilterInterface`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
TYVM @boesing 🙏
